src/mame/f256/tiny_vicky.cpp

```diff
@@ -0,0 +1,613 @@
+    , m_sof_irq_handler(*this)
+    , m_sol_irq_handler(*this)
+{
+}
+
+tiny_vicky_video_device::tiny_vicky_video_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock):
+    tiny_vicky_video_device(mconfig, TINY_VICKY, tag, owner, clock)
+{
+}
+
+rgb_t tiny_vicky_video_device::get_text_lut(uint8_t color_index, bool fg, bool gamma)
+{
+    uint8_t red =   m_iopage0_ptr[(fg ? 0x1800 : 0x1840) + color_index * 4 + 2];
+    uint8_t green = m_iopage0_ptr[(fg ? 0x1800 : 0x1840) + color_index * 4 + 1];
+    uint8_t blue =  m_iopage0_ptr[(fg ? 0x1800 : 0x1840) + color_index * 4 ];
+    if (gamma)
+    {
+        blue = m_iopage0_ptr[blue];
+        green = m_iopage0_ptr[0x400 + green];
+        red = m_iopage0_ptr[0x800 + red];
+    }
+    return rgb_t(red, green, blue);
+}
+
+rgb_t tiny_vicky_video_device::get_lut_value(uint8_t lut_index, uint8_t pix_val, bool gamma)
+{
+    int lutAddress = 0xD000 - 0xC000 + (lut_index * 256 + pix_val) * 4;
+    if (!gamma)
+    {
+        return rgb_t(m_iopage1_ptr[lutAddress + 2], m_iopage1_ptr[lutAddress + 1], m_iopage1_ptr[lutAddress]);
+    }
+    else
+    {
+        return rgb_t(m_iopage0_ptr[0x800 + m_iopage1_ptr[lutAddress + 2]],
+                    m_iopage0_ptr[0x400 + m_iopage1_ptr[lutAddress + 1]],
+                    m_iopage0_ptr[m_iopage1_ptr[lutAddress]]);
+    }
+}
+
+uint32_t tiny_vicky_video_device::screen_update(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect)
+{
+    if (m_running && m_iopage0_ptr)
+    {
+        uint8_t mcr = m_iopage0_ptr[0x1000];
+        uint8_t mcr_h = m_iopage0_ptr[0x1001];
+        m_cursor_counter++;
+        if (m_cursor_counter > m_cursor_flash_rate *2)
+        {
+            m_cursor_counter = 0;
+        }
+        // if MCR=0 or MCR bit 7 is set, then video is disabled
+        if (mcr != 0 && (mcr != 0x80))
+        {
+            // TODO: generate start of frame (SOF) interrupt
+            m_sof_irq_handler(1);
+            uint8_t border_reg = m_iopage0_ptr[0x1004];
+            bool display_border = (border_reg & 0x1) > 0;
+            bool enable_gamma = (mcr & 0x40) > 0;
+            bool enable_graphics = (mcr & 0x4) > 0;
+            uint16_t lines = (mcr_h & 1) == 0 ? 480 : 400;
+
+            uint8_t border_x = 0;
+            uint8_t border_y = 0;
+            rgb_t border_color = rgb_t();
+            if (display_border) {
+                border_x = m_iopage0_ptr[0x1008];
+                border_y = m_iopage0_ptr[0x1009];
+            }
+            uint32_t *topleft = (uint32_t *)bitmap.raw_pixptr(0);
+            for (int y = 0; y < lines; y++)
+            {
+                // Check the Sart of Line registers
+                uint8_t sol_reg = m_iopage0_ptr[0x1018];
+                // 12-bit line
+                uint16_t sol_line = m_iopage0_ptr[0x1018] + ((m_iopage0_ptr[0x101A] & 0xF) << 8);
+                uint32_t *row = topleft + y * 800;
```

Comment:
This isn’t safe use of bitmaps.  You should get a pointer to a row with `&bitmap.pix(y)`.